rclpy/src/rclpy/publisher.cpp

```diff
@@ -75,60 +48,62 @@ publisher_create(
     publisher_ops.qos = pyqos_profile.cast<rmw_qos_profile_t>();
   }
 
-  // Use smart pointer to make sure memory is free'd on error
-  auto deleter = [](rclpy_publisher_t * ptr) {_rclpy_destroy_publisher(ptr);};
-  auto pub = std::unique_ptr<rclpy_publisher_t, decltype(deleter)>(
-    static_cast<rclpy_publisher_t *>(PyMem_Malloc(sizeof(rclpy_publisher_t))),
-    deleter);
-  if (!pub) {
-    throw std::bad_alloc();
-  }
-  pub->publisher = rcl_get_zero_initialized_publisher();
-  pub->node = node;
+  rcl_publisher_ = std::shared_ptr<rcl_publisher_t>(
+    new rcl_publisher_t,
+    [this](rcl_publisher_t * publisher)
+    {
+      if (!node_handle_) {
```

Comment:
@ahcorde how can this ever happen? IIRC the whole purpose of `Handle` is to ensure ordered destruction.

Reply:
maybe there is some wrong with the destructor somewhere else, but If I remove this lines then the test_qos_event will fail

Reply:
Indeed. Default destructors for `Publisher` and `Subscription` are dropping `node_handle_` too early (and thus leaking!). Inverting member fields order or a custom destructor should do.